Makefile

```diff
@@ -1,18 +1,52 @@
+.PHONY: help build/kanidmd build/radiusd test/kanidmd push/kanidmd push/radiusd vendor-prep doc install-tools prep
 
-docker-kanidmd:
-	docker build -f kanidmd/Dockerfile -t kanidm/server:latest .
+IMAGE_BASE ?= kanidm
+IMAGE_VERSION ?= latest
 
-docker-radiusd:
-	docker build -f kanidm_rlm_python/Dockerfile -t kanidm/radius:latest kanidm_rlm_python
+.DEFAULT: help
+help:
+	@fgrep -h "##" $(MAKEFILE_LIST) | fgrep -v fgrep | sed -e 's/\\$$//' | sed -e 's/##/\n\t/'
+
+build/kanidmd:	## build kanidmd images
+build/kanidmd:
+	@docker build -f kanidmd/Dockerfile -t $(IMAGE_BASE)/server:$(IMAGE_VERSION) .
+	@docker build -f kanidmd/Dockerfile \
+		--target alpine \
+		-t $(IMAGE_BASE)/server:$(IMAGE_VERSION)-alpine \
+		.
+
+build/radiusd:	## build radiusd image
+build/radiusd:
+	@docker build -f kanidm_rlm_python/Dockerfile -t $(IMAGE_BASE)/radius:$(IMAGE_VERSION) \
+		kanidm_rlm_python
+
+test/kanidmd:	## test kanidmd
+test/kanidmd:
+	@docker build -f kanidmd/Dockerfile \
+		--target builder \
+		-t $(IMAGE_BASE)/server:$(IMAGE_VERSION)-builder \
+		.
+	@docker run --rm $(IMAGE_BASE)/server:$(IMAGE_VERSION)-builder cargo test
+
+push/kanidmd:	## push kanidmd images
+push/kanidmd:
+	@docker push $(IMAGE_BASE)/server:$(IMAGE_VERSION)
+	@docker push $(IMAGE_BASE)/server:$(IMAGE_VERSION)-alpine
+
+push/radiusd:	## push radiusd image
+push/radiusd:
+	@docker push $(IMAGE_BASE)/radius:$(IMAGE_VERSION)
 
 vendor-prep:
 	cargo vendor
 	tar -czf vendor.tar.gz vendor
 
-doc-local:
+doc: ## build doc local
+doc:
```

Comment:
Is this double line a mistake maybe?

Reply:
No, test it with `make help`. It is for dependencies and documentation the first line.

Reply:
Mind blown. Okay, that's awesome.